Client/Application/BackgroundTask.swift

```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import UIKit
+import XCGLogger
+import Shared
+
+private let log = Logger.browserLogger
+
+// A stateful wrapper around starting/ending UIBackgroundTasks
+class BackgroundTask {
+    let application: UIApplication
+    let name: String?
+    let onExpire: () -> Void
+
+    private(set) private var taskId = UIBackgroundTaskInvalid
+
+    init(application: UIApplication, taskName: String?, onExpire: () -> Void) {
+        self.application = application
+        self.name = taskName
+        self.onExpire = onExpire
+    }
+
+    func begin() {
+        self.taskId = application.beginBackgroundTaskWithName(name) {
```

Comment:
I'm pretty sure this isn't actually a fix for the bug.

I just wrote a Playground that replicates the issue.

```
import Foundation

class Foo {
    func doLater(f: dispatch_block_t) -> Int {
        let succeeds: Int64 = 100000
        let fails: Int64 = 10
        let delay = fails
        let when = dispatch_time(DISPATCH_TIME_NOW, delay)
        let queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)
        dispatch_after(when, queue, f)
        return 99
    }

    func attempt() -> Int {
        print("Attempting...")
        var x: Int = 0
        x = doLater {
            print("X is", x)
        }
        return x
    }
}

let f = Foo()
f.attempt()
```

With `fails` as the time — essentially instant — this prints "X is 0". With anything larger, this prints "X is 99".

So I think the change in iOS 10 is that a backgrounding app sometimes has its background tasks run *immediately*, in which case the task ID has not yet been assigned!